src/workerd/io/trace-common.h

```diff
@@ -220,30 +359,192 @@ struct Log final {
+  kj::Date timestamp;
+
+  LogLevel logLevel;
+  kj::Array<kj::byte> data;
+  Tags tags = nullptr;
+
+  void copyTo(rpc::Trace::LogV2::Builder builder) const;
+  LogV2 clone() const;
 };
 
 struct Exception final {
-  explicit Exception(
-      kj::Date timestamp, kj::String name, kj::String message, kj::Maybe<kj::String> stack);
+  struct Detail {
+    kj::Maybe<kj::Own<Exception>> cause = kj::none;
+    kj::Array<kj::Own<Exception>> errors = nullptr;
+    bool remote = false;
+    bool retryable = false;
+    bool overloaded = false;
+    bool durableObjectReset = false;
+    Tags tags = nullptr;
+
+    Detail clone() const;
+  };
+
+  explicit Exception(kj::Date timestamp,
+      kj::String name,
+      kj::String message,
+      kj::Maybe<kj::String> stack,
+      kj::Maybe<Detail> detail = kj::none);
   Exception(rpc::Trace::Exception::Reader reader);
   Exception(Exception&&) = default;
+  Exception& operator=(Exception&&) = default;
   KJ_DISALLOW_COPY(Exception);
-  ~Exception() noexcept(false) = default;
 
   // Only as accurate as Worker's Date.now(), for Spectre mitigation.
   kj::Date timestamp;
 
   kj::String name;
   kj::String message;
-
   kj::Maybe<kj::String> stack;
+  Detail detail;
+
+  void copyTo(rpc::Trace::Exception::Builder builder) const;
+  Exception clone() const;
+};
+
+struct Subrequest final {
+  using Info = kj::OneOf<FetchEventInfo, JsRpcEventInfo>;
+  explicit Subrequest(uint32_t id, kj::Maybe<Info> info);
+  Subrequest(rpc::Trace::Subrequest::Reader reader);
+  Subrequest(Subrequest&&) = default;
+  Subrequest& operator=(Subrequest&&) = default;
+  KJ_DISALLOW_COPY(Subrequest);
+
+  uint32_t id;
+  kj::Maybe<Info> info;
+
+  void copyTo(rpc::Trace::Subrequest::Builder builder) const;
+  Subrequest clone() const;
+};
+
+struct SubrequestOutcome final {
+  using Info = kj::OneOf<FetchResponseInfo, Tags>;
+  explicit SubrequestOutcome(
+      uint32_t id, kj::Maybe<Info> info, rpc::Trace::Span::SpanOutcome outcome);
+  SubrequestOutcome(rpc::Trace::SubrequestOutcome::Reader reader);
+  SubrequestOutcome(SubrequestOutcome&&) = default;
+  SubrequestOutcome& operator=(SubrequestOutcome&&) = default;
+  KJ_DISALLOW_COPY(SubrequestOutcome);
+
+  uint32_t id;
+  kj::Maybe<Info> info;
+  rpc::Trace::Span::SpanOutcome outcome;
+
+  void copyTo(rpc::Trace::SubrequestOutcome::Builder builder) const;
+  SubrequestOutcome clone() const;
+};
+
+struct SpanEvent final {
+  using Outcome = rpc::Trace::Span::SpanOutcome;
+  using Info = kj::OneOf<FetchResponseInfo, ActorFlushInfo, Tags>;
+  explicit SpanEvent(uint32_t id,
+      uint32_t parentId,
+      rpc::Trace::Span::SpanOutcome outcome,
+      bool transactional = false,
+      kj::Maybe<Info> maybeInfo = kj::none,
+      Tags tags = nullptr);
+  SpanEvent(rpc::Trace::Span::Reader reader);
+  SpanEvent(SpanEvent&&) = default;
+  SpanEvent& operator=(SpanEvent&&) = default;
+  KJ_DISALLOW_COPY(SpanEvent);
+
+  uint32_t id;
+  uint32_t parent;
```

Comment:
Maybe `parentId`, instead of parent is much verbose?

Reply:
I generally prefer just the `parent` here.